cpp/benchmarks/join/join_benchmark.cu

```diff
@@ -0,0 +1,120 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <benchmark/benchmark.h>
+#include <vector>
+#include <cudf/cudf.h>
+#include <utilities/error_utils.hpp>
+#include <tests/utilities/column_wrapper.cuh>
+
+#include "generate_input_tables.cuh"
+
+
+template<typename key_type, typename payload_type>
+static void join_benchmark(benchmark::State& state)
+{
+    const gdf_size_type build_table_size {(gdf_size_type) state.range(0)};
+    const gdf_size_type probe_table_size {(gdf_size_type) state.range(1)};
+    const gdf_size_type rand_max_val {build_table_size * 2};
+    const double selectivity = 0.3;
+    const bool is_build_table_key_unique = true;
+
+    // Generate build and probe tables
+
+    cudf::test::column_wrapper<key_type> build_key_column(build_table_size);
+    cudf::test::column_wrapper<key_type> probe_key_column(probe_table_size);
+
+    generate_input_tables<key_type, gdf_size_type>(
+        (key_type *)build_key_column.get()->data, build_table_size,
+        (key_type *)probe_key_column.get()->data, probe_table_size,
+        selectivity, rand_max_val, is_build_table_key_unique
+    );
+
+    cudf::test::column_wrapper<payload_type> build_payload_column(
+        build_table_size,
+        [] (gdf_index_type row_index) {
+            return row_index;
+        }
+    );
+
+    cudf::test::column_wrapper<payload_type> probe_payload_column(
+        probe_table_size,
+        [] (gdf_index_type row_index) {
+            return row_index;
+        }
+    );
+
+    CUDA_TRY(cudaGetLastError());
+    CUDA_TRY(cudaDeviceSynchronize());
+
+    std::vector<gdf_column*> build_table {build_key_column.get(), build_payload_column.get()};
+    std::vector<gdf_column*> probe_table {probe_key_column.get(), probe_payload_column.get()};
+
+    // Setup join parameters and result table
+
+    gdf_context ctxt = {
+        0,                     // input data is not sorted
+        gdf_method::GDF_HASH   // hash based join
+    };
+
+    int columns_to_join[] = {0};
+
+    const int nresult_cols = build_table.size() + probe_table.size() - 1;
+
+    std::vector<gdf_column> result_table(nresult_cols);
+    std::vector<gdf_column *> col_ptrs(nresult_cols);
+
+    for (size_t icol = 0; icol < result_table.size(); icol++) {
+        col_ptrs[icol] = &result_table[icol];
+    }
+
+    // Benchmark the inner join operation
+
+    CUDA_TRY(cudaDeviceSynchronize());
+
+    for (auto _ : state) {
+        CUDF_TRY(gdf_inner_join(
+            probe_table.data(), 2, columns_to_join,
+            build_table.data(), 2, columns_to_join,
+            1, nresult_cols, col_ptrs.data(),
+            nullptr, nullptr, &ctxt
+        ));
+
+        CUDA_TRY(cudaDeviceSynchronize());
```

Comment:
Definitely don't want to cudaDeviceSynchronize inside the benchmark loop -- this will penalize the benchmark.

Reply:
Suppose there are async operations inside the join computation. If we do not have a `cudaDeviceSynchronize` here, is it possible that we record the stop time but the actual computation hasn't completed yet?

Reply:
We should call cudaDeviceSynchronize() on the last iteration, just before the timer stops. There must be a way to do this with google benchmark...

Reply:
I didn't find any documentation on how to test for the last iteration in Google benchmark.

Reply:
@gaohao95 you'll find this relevant: https://github.com/google/benchmark#manual-timing

@hummingtree is working on creating an abstraction to ease timing kernel execution without needing to synchronize for every iteration.

Reply:
Resolved by using manual timing and cuda events in https://github.com/rapidsai/cudf/pull/2105/commits/1e2d1582b627526e6a0c37e65aa746bbe8e0c608.